processing/src/main/java/org/apache/druid/segment/DimensionHandlerUtils.java

```diff
@@ -305,17 +305,29 @@ private static ColumnCapabilities getEffectiveCapabilities(
         originalCapabilities == null || ValueType.COMPLEX.equals(originalCapabilities.getType());
 
     if (type == ValueType.STRING) {
-      if (!forceSingleValue && effectiveCapabilites.hasMultipleValues().isMaybeTrue()) {
-        return strategyFactory.makeMultiValueDimensionProcessor(
-            effectiveCapabilites,
-            selectorFactory.makeMultiValueDimensionSelector(dimensionSpec)
-        );
-      } else {
-        return strategyFactory.makeSingleValueDimensionProcessor(
-            effectiveCapabilites,
-            selectorFactory.makeSingleValueDimensionSelector(dimensionSpec)
-        );
+      if (!forceSingleValue) {
+        // if column is not dictionary encoded (and not non-existent or complex), use object selector
+        if (effectiveCapabilites.isDictionaryEncoded().isFalse() ||
+            effectiveCapabilites.areDictionaryValuesUnique().isFalse()
```

Comment:
by regular string selector do you mean dimension selector? The answer is yes we _could_, but if the dictionary values are not unique, i came to the conclusion that they aren't especially useful since in all the cases I can think of you can't use them as is and you're going to have to lookup the string values anyway (e.g. filter matching and vector group by).

I was trying to provide an avenue to avoid the extra steps with expression virtual columns, and allow an object selector to be created directly since the string is the useable part in this case, and the dictionary ids just feel like an obstruction. But maybe this is overly broad and there are cases where we would want it to act like a dictionary encoded column for some reason that I have missed?

Reply:
I think your reasoning makes sense, and it sounds like your assumption is that when selecting string-typed things that aren't dictionary-encoded, we should prefer the object selector vs. one of the DimensionVectorSelectors, because that'll incur less overhead.

This logic should be spelled out somewhere though. IMO javadocs on the appropriate methods in VectorColumnSelectorFactory is the best place. They should say when callers should prefer using the object selector vs. the more specifically-typed selectors.

Reply:
I think I've got this covered better now